### PR TITLE
Jetty 10.0.x websocket refactor

### PR DESCRIPTION
**Pull request by gregw**
A major refactor of websocket to be based around a core implementation

### CODE REVIEW

**Comment by gregw** [comment]
This PR is for the technical review before merging back to 10.0.x

File: jetty-client/src/main/java/org/eclipse/jetty/client/HttpClient.java [hunk 1]
```diff
@@ -212,8 +213,8 @@ protected void doStart() throws Exception
             QueuedThreadPool threadPool = new QueuedThreadPool();
             threadPool.setName(name);
             executor = threadPool;
+            addBean(executor);
```
Comment (joakime) [inline jetty-client/src/main/java/org/eclipse/jetty/client/HttpClient.java]: Looks like a useful change to backport to jetty-9.4.x

Comment (joakime) [reply jetty-client/src/main/java/org/eclipse/jetty/client/HttpClient.java]: Along with the `updateBean()` in the `setExecutor(Executor)` call below

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java [hunk 2]
```diff
@@ -795,6 +795,19 @@ public URI toURI() throws URISyntaxException
         return new URI(_scheme,null,_host,_port,_path,_query==null?null:UrlEncoded.decodeString(_query),_fragment);
     }
 
+    /* ------------------------------------------------------------ */
+    public URI asURI()
```
Comment (joakime) [inline jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java]: Do we really need this?  isn't the existing `toURI()` sufficient?

File: jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/UriTemplatePathSpec.java [hunk 3]
```diff
@@ -321,7 +322,8 @@ private boolean isValidBasicLiteralCodepoint(int codepoint)
             int groupCount = matcher.groupCount();
             for (int i = 1; i <= groupCount; i++)
             {
-                ret.put(this.variables[i - 1],matcher.group(i));
+                String value = UrlEncoded.decodeString(matcher.group(i));
```
Comment (joakime) [inline jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/UriTemplatePathSpec.java]: Doesn't `UrlEncoded.decodeString()` require a charset?

File: jetty-http/src/test/java/org/eclipse/jetty/http/HttpURIParseTest.java [hunk 4]
```diff
@@ -109,8 +109,8 @@
         Arguments.of("file:///path/info","file",null,null,"/path/info",null,null,null),
         Arguments.of("file:/path/info","file",null,null,"/path/info",null,null,null),
         
-        // Bad URI (no scheme, no host, no path) 
-        Arguments.of("//",null,null,null,null,null,null,null),
+        // Bad URI (no scheme, no host, no path)
+        // Arguments.of("//",null,null,null,null,null,null,null),
```
Comment (joakime) [inline jetty-http/src/test/java/org/eclipse/jetty/http/HttpURIParseTest.java]: Move this one to the "HttpURI testcase of "Bad" URIs?

File: jetty-io/src/main/java/org/eclipse/jetty/io/ChannelEndPoint.java [hunk 5]
```diff
@@ -229,29 +229,29 @@ public int fill(ByteBuffer buffer) throws IOException
...
 
         int pos=BufferUtil.flipToFill(buffer);
+        int filled;
         try
         {
-            int filled = _channel.read(buffer);
-            if (LOG.isDebugEnabled()) // Avoid boxing of variable 'filled'
-                LOG.debug("filled {} {}", filled, this);
-
+            filled = _channel.read(buffer);
```
Comment (joakime) [inline jetty-io/src/main/java/org/eclipse/jetty/io/ChannelEndPoint.java]: Are the changes in `ChannelEndPoint` worth backporting to jetty-9.4.x too?

File: jetty-util/src/main/java/org/eclipse/jetty/util/BufferUtil.java [hunk 6]
```diff
@@ -387,6 +403,29 @@ public static void append(ByteBuffer to, byte[] b, int off, int len) throws Buff
...
+     * @param b bytes to append
+     * @throws BufferOverflowException if unable to append buffer due to space limits
+     */
+    public static void append(ByteBuffer to, byte[] b) throws BufferOverflowException
+    {
+        append(to,b,0,b.length);
+    }
+    
+    /* ------------------------------------------------------------ */
+    /** Append bytes to a buffer.
```
Comment (joakime) [inline jetty-util/src/main/java/org/eclipse/jetty/util/BufferUtil.java]: Seems like a copy/paste flub.
Perhaps it should be "Append String UTF-8 bytes to a buffer" ?

File: jetty-util/src/main/java/org/eclipse/jetty/util/IteratingCallback.java [hunk 7]
```diff
@@ -403,7 +403,7 @@ public void failed(Throwable x)
 
     public void close()
     {
-        boolean failure=false;
+        String failure=null;
```
Comment (joakime) [inline jetty-util/src/main/java/org/eclipse/jetty/util/IteratingCallback.java]: The changes in `IteratingCallback` seems like a candidate for jetty-9.4.x backport.

File: jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/UriTemplatePathSpec.java (hunk 3 above)
Comment (gregw) [reply jetty-http/src/main/java/org/eclipse/jetty/http/pathmap/UriTemplatePathSpec.java]: It uses the default (UTF-8, but configurable) in UrlEncoded.

**Pull request merged**